src/decoder/symphonia.rs

```diff
@@ -273,72 +282,60 @@ impl std::error::Error for SeekError {
+            self.time_base
+                .ok_or(source::SeekError::SymphoniaDecoder(SeekError::Refining(
+                    "time base not available",
+                )))?
+                .calc_time(seek_res.required_ts.saturating_sub(seek_res.actual_ts)),
+        )
+        .as_secs_f32()
+            * self.sample_rate() as f32
+            * self.channels() as f32)
+            .ceil() as usize;
+
+        // Re-align the seek position to the first channel.
+        samples_to_skip -= samples_to_skip % self.channels() as usize;
+
+        // Skip ahead to the precise position.
+        for _ in 0..samples_to_skip {
+            self.next();
         }
 
-        let decoded = decoded.map_err(SeekError::Decoding)?;
-        decoded.spec().clone_into(&mut self.spec);
-        self.buffer = SymphoniaDecoder::get_buffer(decoded, &self.spec);
-        self.current_span_offset = samples_to_pass as usize * self.channels() as usize;
         Ok(())
     }
 }
 
-fn skip_back_a_tiny_bit(
-    Time {
-        mut seconds,
-        mut frac,
-    }: Time,
-) -> Time {
-    frac -= 0.0001;
-    if frac < 0.0 {
-        seconds = seconds.saturating_sub(1);
-        frac = 1.0 - frac;
-    }
-    Time { seconds, frac }
-}
-
-fn time_to_duration(time: Time) -> Duration {
-    Duration::new(
-        time.seconds,
-        if time.frac > 0.0 {
-            (1f64 / time.frac) as u32
-        } else {
-            0
-        },
-    )
-}
-
 impl Iterator for SymphoniaDecoder {
     type Item = DecoderSample;
 
-    #[inline]
     fn next(&mut self) -> Option<Self::Item> {
         if self.current_span_offset >= self.buffer.len() {
-            let packet = self.format.next_packet().ok()?;
-            let mut decoded = self.decoder.decode(&packet);
-            for _ in 0..MAX_DECODE_RETRIES {
+            let mut decode_errors = 0;
+            let decoded = loop {
+                let packet = self.format.next_packet().ok()?;
+                let decoded = self.decoder.decode(&packet);
                 if decoded.is_err() {
-                    let packet = self.format.next_packet().ok()?;
-                    decoded = self.decoder.decode(&packet);
+                    decode_errors += 1;
+                    if decode_errors > MAX_DECODE_RETRIES {
+                        return None;
+                    } else {
+                        continue;
+                    }
                 }
-            }
-            let decoded = decoded.ok()?;
+
+                // Loop until we get a packet with audio frames. This is necessary because some
+                // formats can have packets with only metadata, particularly when rewinding, in
+                // which case the iterator would otherwise end with `None`.
+                // Note: checking `decoded.frames()` is more reliable than `packet.dur()`, which
+                // can resturn non-zero durations for packets without audio frames.
+                let decoded = decoded.ok()?;
```

Comment:
this `.ok()?` should be an `.expect("if above handles Err variant")` or the `if decoded.is_err` could become a match returning the `Ok` value.

Reply:
Good catch.